ddsketch/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+            #[allow(clippy::needless_return)]
+            return 0;
+        }
+        // Bucket higher than the stored range
+        else if self.offset + self.bins.len() as i32 <= bin_index {
+            let bin_range_size = bin_index - self.offset + 1; // Number of bucket to store
+
+            if bin_range_size > self.max_size {
+                self.collapse_low_bins(bin_range_size - self.max_size);
+            }
+            debug_assert!(self.bins.len() as i32 <= self.max_size);
+
+            let store_index = bin_index - self.offset;
+            for _ in 0..(store_index - self.bins.len() as i32 + 1) {
+                self.bins.push_back(0.0);
+            }
+            return store_index as usize;
+        }
+        // Bucket within the stored range
+        else {
+            return (bin_index - self.offset) as usize;
+        }
+    }
+
+    /// Collapse the `bin_number` lowest bins
+    fn collapse_low_bins(&mut self, bin_number: i32) {
+        let mut count = 0.0;
+        for _ in 0..bin_number {
+            count += self.bins.pop_front().unwrap_or(0.0);
+        }
+        if let Some(lowest_bin) = self.bins.front_mut() {
+            *lowest_bin += count;
+        } else {
+            self.bins.push_front(count);
+        }
+        self.offset += bin_number;
+    }
+}
+
+impl Default for LowCollapsingDenseStore {
+    fn default() -> Self {
+        Self::new(2048).unwrap()
+    }
+}
+
+/// Logarithmic mapping of bucket index to value
+#[derive(Debug, Clone, Copy)]
+struct LogMapping {
+    gamma: f64,
+    multiplier: f64,
+    min_indexable_value: f64,
+    index_offset: f64,
+}
+
+impl LogMapping {
+    fn new(gamma: f64, offset: f64) -> Option<Self> {
+        if gamma <= 1.0 {
+            return None;
+        }
+        let multiplier = Self::multiplier_from_gamma(gamma);
+        Some(Self {
+            gamma,
+            multiplier,
+            min_indexable_value: max(
+                // So that the value representing the lowest bucket is >= std::f64::MIN_POSITIVE
+                std::f64::MIN_POSITIVE * gamma,
+                // Minimum value so that index >= i32::MIN
+                ((i32::MIN as f64 - offset) / multiplier + 1.0).exp(),
+            )?,
+            index_offset: offset,
+        })
+    }
+
+    /// Returns the multiplier used to convert ln to base-gamma logarithm
+    fn multiplier_from_gamma(gamma: f64) -> f64 {
+        1.0 / gamma.ln()
+    }
+
+    /// Returns the relative accuracy guaranteed by the mapping
+    fn relative_accuracy(&self) -> f64 {
+        1.0 - 2.0 / (1.0 + self.gamma)
+    }
+
+    /// Returns the index of the bucket containing `value`
+    fn index(&self, value: f64) -> i32 {
+        (value.ln() * self.multiplier + self.index_offset).floor() as i32
+    }
+
+    /// Returns the value representing the bucket at `index`
+    fn value(&self, index: i32) -> f64 {
+        ((index as f64 - self.index_offset) / self.multiplier).exp()
```

Comment:
Just curious, is there no need to be concerned with floating point math accuracy for ddsketch? Are we safe to not use an arbitrary-precision floating point library like BigDecimal?